atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/NameComponentDescription.java

```diff
@@ -20,71 +20,79 @@
+        private UniformRowNamePartitioner uniformPartitioner;
+        private ExplicitRowNamePartitioner explicitPartitioner;
+        private LogSafety logSafety = LogSafety.UNSAFE;
+
+        public Builder componentName(String name) {
+            this.componentName = name;
+            return this;
+        }
 
-    public NameComponentDescription(String componentName, ValueType type) {
-        this(componentName, type, ValueByteOrder.ASCENDING, new UniformRowNamePartitioner(type), null);
-    }
+        public Builder type(ValueType valueType) {
+            this.type = valueType;
+            return this;
+        }
 
-    @Deprecated
-    public NameComponentDescription(String componentName, ValueType type, boolean reverseOrder) {
-        this.componentName = componentName;
-        this.type = type;
-        this.order = reverseOrder ? ValueByteOrder.DESCENDING : ValueByteOrder.ASCENDING;
-        this.uniformPartitioner = new UniformRowNamePartitioner(type);
-        this.explicitPartitioner = null;
-        this.logSafety = LogSafety.UNSAFE;
-    }
+        public Builder byteOrder(ValueByteOrder byteOrder) {
+            this.order = byteOrder;
+            return this;
+        }
 
-    public NameComponentDescription(String componentName,
-            ValueType type,
-            ValueByteOrder order) {
-        this(componentName, type, order, LogSafety.UNSAFE);
-    }
+        public Builder uniformRowNamePartitioner(UniformRowNamePartitioner partitioner) {
+            this.uniformPartitioner = partitioner;
+            return this;
+        }
 
-    public NameComponentDescription(String componentName,
-            ValueType type,
-            ValueByteOrder order,
-            LogSafety logSafety) {
-        this(componentName, type, order, new UniformRowNamePartitioner(type), null, logSafety);
-    }
+        public Builder explicitRowNamePartitioner(ExplicitRowNamePartitioner partitioner) {
+            this.explicitPartitioner = partitioner;
+            return this;
+        }
 
-    public NameComponentDescription(String componentName,
-                                    ValueType type,
-                                    ValueByteOrder order,
-                                    UniformRowNamePartitioner uniform,
-                                    ExplicitRowNamePartitioner explicit) {
-        this(componentName, type, order, uniform, explicit, LogSafety.UNSAFE);
+        public Builder logSafety(LogSafety safety) {
+            this.logSafety = safety;
+            return this;
+        }
+
+        public NameComponentDescription build() {
+            Validate.notNull(componentName, "componentName must be set when building a NameComponentDescription");
+            Validate.notNull(type, "type must be set when building a NameComponentDescription");
+
+            if (uniformPartitioner == null) {
+                uniformPartitioner = new UniformRowNamePartitioner(type);
+            }
```

Comment:
I'm a little concerned this might not be safe. It seems e.g. from `withPartitioners` that having explicitly null `uniformPartitioner` is allowed (though that still works).

So I might do something like

```java
  new NameComponentDescription.Builder()
          .componentName("foo")
          .valueType(ValueType.BLOB)
          .uniformPartitioner(null)
          .explicitPartitioner(somePartitioner).build();
```

and the NCD created will have a uniform partitioner - but a call via the constructor with the same arguments won't.

Looks like the `uniformRowNamePartitioner` and `explicitRowNamePartitioner` are only called by test methods at this time, and people shouldn't be building `NameComponentDescription`s directly, so this won't cause any user-facing issues, but might be nice to address this.